modules/parachains/src/extension.rs

```diff
@@ -72,15 +72,34 @@ macro_rules! declare_bridge_reject_obsolete_parachain_header {
 							ref parachains,
 							..
 						}) if parachains.len() == 1 => {
-							let parachain = parachains.get(0).expect("verified by match condition; qed");
+							let (parachain, parachain_head_hash) = parachains.get(0).expect("verified by match condition; qed");
 
 							let bundled_relay_block_number = at_relay_block.0;
 
 							let best_parachain_head = $crate::BestParaHeads::<$runtime, $instance>::get(parachain);
+
 							match best_parachain_head {
 								Some(best_parachain_head) if best_parachain_head.at_relay_block_number
-									>= bundled_relay_block_number =>
-										sp_runtime::transaction_validity::InvalidTransaction::Stale.into(),
+									>= bundled_relay_block_number => {
+									log::trace!(
+										target: $crate::LOG_TARGET,
+										"Rejecting obsolete parachain-head {:?} transaction: bundled relay block number: \
+										{:?} best relay block number: {:?}",
+										parachain,
+										bundled_relay_block_number,
+										best_parachain_head.at_relay_block_number,
+									);
+									sp_runtime::transaction_validity::InvalidTransaction::Stale.into()
+								}
+								Some(best_parachain_head) if best_parachain_head.head_hash == *parachain_head_hash => {
+									log::trace!(
+										target: $crate::LOG_TARGET,
+										"Rejecting obsolete parachain-head {:?} transaction: head hash {:?}",
+										parachain,
+										best_parachain_head.head_hash,
+									);
+									sp_runtime::transaction_validity::InvalidTransaction::Stale.into()
+								}
```

Comment:
This validation looks very similar with the one [here](https://github.com/paritytech/parity-bridges-common/blob/master/modules/parachains/src/lib.rs#L386-L420). 2 questions:
1. The equality case for `at_relay_block_number` seems to be treated differently. Here it is treated as invalid, in `update_parachain_head()` it's treated as valid. Is this expected ?
2. Would it make sense to do the same validation in both places ? In this case would it make sense to deduplicate the code ?

Reply:
1. Imo in both cases `at_relay_block_number` leads to error, no (I maybe missing something - sorry)?

```rust
fn update_parachain_head() {
    ...
    match stored_best_head {
    	Some(stored_best_head) if stored_best_head.at_relay_block_number <= updated_at_relay_block_number => {
            if updated_head_hash == stored_best_head.head_hash {
                return Err(());
            }

            // else: pass
        }
    ...
}

fn validate() {
	...
	match best_parachain_head {
		Some(best_parachain_head) if best_parachain_head.at_relay_block_number
			>= bundled_relay_block_number => Err(()),
		Some(best_parachain_head) if best_parachain_head.head_hash == *parachain_head_hash => Err(())
	...
}
```

A few notes on that: the other condition (matching hashes) is actually a separate condition, so maybe in the extension it is expressed better, with a dedicated match branch. In general, we shall ignore tx if it supplies head updated at `<= stored_best_head.at_relay_block_number` and we shall ignore if it has matching hashes. In practice we can't have two blocks with the same hash (crypto-safe hashing) - there's a lot that will break if we'll see it, so second condition may be checked separately or not.

2. Yeah - if there's a better way to do that, go ahead! Thank you!

Reply:
> 1. Imo in both cases `at_relay_block_number` leads to error, no (I maybe missing something - sorry)?

From what I understand, if `stored_best_head.at_relay_block_number == updated_at_relay_block_number` and the hashes are different, for `update_parachain_head()` it will go to ` // else: pass`, while the extension will trigger an error. So if I understand correctly this case is treated differently now, even if in practice we shouldn't have to blocks with the same hash.

Anyway, I just wanted to make sure if we should have the same check in both places.